### PR TITLE
BB2-2919:  Update Documentation for Quicksight Terraforming

### PR DESCRIPTION
**Pull request by dtisza1**
**JIRA Ticket:**
[BB2-2919](https://jira.cms.gov/browse/BB2-2919)

**User Story or Bug Summary:**

After the Quicksight terraforming work is complete, developers need updated documentation on how to add new metrics to the Quicksight analyses and dashboards.

### What Does This PR Do?

- Updates are made to the main Terraform related README: `insights/terraform/projects/bb2/README.md`
- Created a more detailed README for the `quicksight` service under: `insights/terraform/projects/bb2/services/quicksight/README.md`
- Updates are made to the README for the `update_athena_metric_tables` lambda under: `services/analytics/modules/lambda/update_athena_metric_tables/README.md`

This updates the documentation related to recent work performed for [BB2-2905](https://jira.cms.gov/browse/BB2-2905) and [BB2-2963](https://jira.cms.gov/browse/BB2-2963) Jira stories.

- Fix for 0's initially showing for the Dynamic field after a new TF APPLY in `services/quicksight/modules/quicksight-analysis-prod-applications/main.tf`.

###  What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: insights/terraform/projects/bb2/README.md [hunk 1]
```diff
@@ -12,43 +12,65 @@ The BB2 specific components are setup under the following service areas:
 
 - `services/analytics`:
   - Includes components for the lambda and event scheduler that update the intermediate reporting tables via Athena and for usage in QuickSight.
+    - For more details, see this module [README.md](services/analytics/modules/lambda/update_athena_metric_tables/README.md).
+    - Is updated when adding new metrics (except calculated fields).
```
Comment (meliGuzman) [inline insights/terraform/projects/bb2/README.md:16]: @dtisza1  Hmm not clear on line 16. Are you saying that this section of code should be modified when adding new metrics?

File: insights/terraform/projects/bb2/README.md [hunk 2]
```diff
@@ -12,43 +12,65 @@ The BB2 specific components are setup under the following service areas:
...
+Normally when terraforming resources, the `terraform plan` after a `terraform apply`, will show the following message: "No changes. Your infrastructure matches the configuration."
+
+However, due to several issues when terraforming Quicksight Analyses, these will always show some types of expected changes. These are described in the following README document: [services/quicksight/README.md](services/quicksight/README.md).
+
+
+## Documentation for Athena & Quicksight Metrics
+
+There is more detailed documention located under the `update_athena_metric_tables` lambda module directory. This contains documentation related to updating the intermediate tables utilized by the datasets. The datasets are used by the analyses that are used to publish the dashboards.
+
+For more details, see this [services/analytics/modules/lambda/update_athena_metric_tables/README.md](services/analytics/modules/lambda/update_athena_metric_tables/README.md).
```
Comment (meliGuzman) [inline insights/terraform/projects/bb2/README.md:76]: @dtisza1 I think it would be a good idea to add my finding of the bug with adding boolean metrics inside the readme under update_athena_metric_tables. I can do that if you'd like!

File: insights/terraform/projects/bb2/services/quicksight/README.md [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+## Summary of Analysis Terraform Issues
+
+The following is a description of some issues related to the plan changes **that always show** in both analyses plan outputs. Even after a successful terraform APPLY.
+
+- Some types of `height` values are set to 0 internally; however, setting this in the TF code produces an error during the apply, with a message that the value must be greater than 0.
+- Some types of `height` values do not change (Ex. height=40 for row_header_style).
+- Some `text_wrap = "WRAP"` values do not change.
+- The `column_configurations` shown will flip positions, when matching them in the TF code and after a new apply.
+- The theme setting does not work.
+
```
Comment (meliGuzman) [inline insights/terraform/projects/bb2/services/quicksight/README.md:607]: @dtisza1 i think there's also something about the dimension / measure that pops up as well

File: insights/terraform/projects/bb2/services/quicksight/modules/quicksight-analysis-prod-applications/main.tf [hunk 4]
```diff
@@ -73,7 +73,7 @@ resource "aws_quicksight_analysis" "quicksight_analysis_prod_applications" {
...
...count},\n$${DYNAMICFIELD} = 'AUTH-FLOW Auth Matched Returning Enrollee Real', {calc_app_authentication_matched_returning_bene_real_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW Auth Matched Returning Enrollee Synthetic', {calc_app_authentication_matched_returning_bene_synthetic_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW SLS CallBack OK Real', {calc_app_sls_callback_ok_real_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW SLS CallBack OK Synthetic', {calc_app_sls_callback_ok_synthetic_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW SLS CallBack FAIL', {calc_app_sls_callback_fail_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW Approval View GET OK Real', {calc_app_approval_view_get_ok_real_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW Approval View GET OK Synthetic', {calc_app_approval_view_get_ok_synthetic_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW Approval View GET FAIL', {calc_app_approval_view_get_fail_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW Approval View POST OK Real', {calc_app_approval_view_post_ok_real_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW Approval View POST OK Synthetic', {calc_app_approval_view_post_ok_synthetic_count},\n$${DYNAMICFIELD} = 'AUTH-FLOW Approval View POST FAIL', {calc_app_approval_view_post_fail_count},\n 0)"
```
Comment (meliGuzman) [inline insights/terraform/projects/bb2/services/quicksight/modules/quicksight-analysis-prod-applications/main.tf:76]: @dtisza1 so that 0 at the very end is what makes the values default to 0. It looks like you removed it and replace it with {calc_app_grant_and_archived_real_bene_deduped_count}. Is that correct? Is the reason to default to this metric if none selected?

File: insights/terraform/projects/bb2/README.md (hunk 1 above)
Comment (dtisza1) [reply insights/terraform/projects/bb2/README.md:16]: > @dtisza1 Hmm not clear on line 16. Are you saying that this section of code should be modified when adding new metrics?

@meliGuzman  Was trying to denote which of the service areas are related to adding new metrics.  Maybe should remove that from here, since it is documented elsewhere. For example, the lambda templates need updated under that area for new metrics.  Does removing or rewording sound good to you?

File: insights/terraform/projects/bb2/README.md (hunk 2 above)
Comment (dtisza1) [reply insights/terraform/projects/bb2/README.md:76]: > @dtisza1 I think it would be a good idea to add my finding of the bug with adding boolean metrics inside the readme under update_athena_metric_tables. I can do that if you'd like!

I'm going to work on adding to that README next. I'll include something about that. Probably not good to update the code to handle boolean values now. Not wanting to break anything :-)    Might be good to put in a backlog Jira story about it?

File: insights/terraform/projects/bb2/services/quicksight/modules/quicksight-analysis-prod-applications/main.tf (hunk 4 above)
Comment (dtisza1) [reply insights/terraform/projects/bb2/services/quicksight/modules/quicksight-analysis-prod-applications/main.tf:76]: > @dtisza1 so that 0 at the very end is what makes the values default to 0. It looks like you removed it and replace it with {calc_app_grant_and_archived_real_bene_deduped_count}. Is that correct? Is the reason to default to this metric if none selected?

That's right. I'm not sure why, but the original (non-TF) analysis was defaulting to the 1st value, so that 0 was never an issue. After the terraforming, it is initially using that 0 value instead of the 1st dynamic field. If that makes sense?

File: insights/terraform/projects/bb2/README.md (hunk 1 above)
Comment (meliGuzman) [reply insights/terraform/projects/bb2/README.md:16]: @dtisza1 i think either or works so maybe reword this to "when adding new metrics except for calculated field services/analytics would need to be updated"

File: insights/terraform/projects/bb2/README.md (hunk 2 above)
Comment (meliGuzman) [reply insights/terraform/projects/bb2/README.md:76]: @dtisza1  I agree! I can open a ticket on that bug

Comment (dtisza1) [reply insights/terraform/projects/bb2/README.md:76]: @meliGuzman TY! That would work out great with a backlog story for that bug.

File: insights/terraform/projects/bb2/services/quicksight/README.md (hunk 3 above)
Comment (dtisza1) [reply insights/terraform/projects/bb2/services/quicksight/README.md:626]: > @dtisza1 i think there's also something about the dimension / measure that pops up as well

@meliGuzman I added some more info for that `column_configurations` . Does that read better?

File: insights/terraform/projects/bb2/README.md (hunk 2 above)
Comment (meliGuzman) [reply insights/terraform/projects/bb2/README.md:76]: @dtisza1 bug ticket created! https://jira.cms.gov/browse/BB2-3133

**Review by meliGuzman — APPROVED** [review]

**Review by dondevun — APPROVED** [review]
